File: maki-code-index/src/java.rs
use tree_sitter::Node;

use crate::common::{
    ChildKind, FIELD_TRUNCATE_THRESHOLD, LanguageExtractor, Section, SkeletonEntry, compact_ws,
    find_child, line_range, node_text, prefixed,
};

pub(crate) struct JavaExtractor;

impl JavaExtractor {
    fn type_list_text(&self, parent: Node, source: &[u8]) -> String {
        let Some(tl) = find_child(parent, "type_list") else {
            return node_text(parent, source)
                .trim_start_matches("extends")
                .trim_start_matches("implements")
                .trim()
                .to_string();
        };
        node_text(tl, source).to_string()
    }

    fn implements_clause(&self, node: Node, source: &[u8]) -> String {
        node.child_by_field_name("interfaces")
            .map(|n| format!(" implements {}", self.type_list_text(n, source)))
            .unwrap_or_default()
    }

    fn extract_import(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let text = node_text(node, source);
        let cleaned = text
            .strip_prefix("import ")
            .unwrap_or(text)
            .trim_end_matches(';')
            .trim();
        let paths = vec![
            cleaned
                .split(self.import_separator())
                .map(String::from)
                .collect(),
        ];
        Some(SkeletonEntry::new_import(node, paths))
    }

    fn extract_package(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let text = node_text(node, source);
        let cleaned = text
            .strip_prefix("package ")
            .unwrap_or(text)
            .trim_end_matches(';')
            .trim()
            .to_string();
        Some(SkeletonEntry::new(Section::Module, node, cleaned))
    }

    fn modifiers_text(&self, node: Node, source: &[u8]) -> String {
        let Some(mods) = find_child(node, "modifiers") else {
            return String::new();
        };
        let mut annotations = Vec::new();
        let mut keywords = Vec::new();
        let mut cursor = mods.walk();
        for child in mods.children(&mut cursor) {
            match child.kind() {
                "marker_annotation" | "annotation" => {
                    annotations.push(node_text(child, source));
                }
                _ => {
                    let text = node_text(child, source);
                    if matches!(
                        text,
                        "public"
                            | "private"
                            | "protected"
                            | "static"
                            | "final"
                            | "abstract"
                            | "default"
                            | "synchronized"
                    ) {
                        keywords.push(text);
                    }
                }
            }
        }
        annotations.extend(keywords);
        annotations.join(" ")
    }

    fn extract_class(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let mods = self.modifiers_text(node, source);
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))?;
        let type_params = find_child(node, "type_parameters")
            .map(|n| node_text(n, source))
            .unwrap_or("");
        let superclass = node
            .child_by_field_name("superclass")
            .and_then(|n| find_child(n, "type_identifier").or(Some(n)))
            .map(|n| format!(" extends {}", node_text(n, source)))
            .unwrap_or_default();
        let interfaces = self.implements_clause(node, source);

        let label = prefixed(
            &mods,
            format_args!("class {name}{type_params}{superclass}{interfaces}"),
        );

        let children = self.extract_class_body(node, source);
        Some(SkeletonEntry::new(Section::Class, node, label).with_children(children))
    }

    fn extract_class_body(&self, node: Node, source: &[u8]) -> Vec<String> {
        let Some(body) = node.child_by_field_name("body") else {
            return Vec::new();
        };
        let mut members = Vec::new();
        let mut field_count = 0usize;
        let mut cursor = body.walk();
        for child in body.children(&mut cursor) {
            match child.kind() {
                "method_declaration" | "constructor_declaration" => {
                    let sig = self.method_signature(child, source);
                    let lr =
                        line_range(child.start_position().row + 1, child.end_position().row + 1);
                    members.push(format!("{sig} {lr}"));
                }
                "field_declaration" => {
                    field_count += 1;
                    if field_count <= FIELD_TRUNCATE_THRESHOLD {
                        let text = self.field_text(child, source);
                        let lr = line_range(
                            child.start_position().row + 1,
                            child.end_position().row + 1,
                        );
                        members.push(format!("{text} {lr}"));
                    }
                }
                _ => {}
            }
        }
        if field_count > FIELD_TRUNCATE_THRESHOLD {
            members.push("...".into());
        }
        members
    }

    fn method_signature(&self, node: Node, source: &[u8]) -> String {
        let mods = self.modifiers_text(node, source);
        let ret = node
            .child_by_field_name("type")
            .map(|n| node_text(n, source))
            .unwrap_or("");
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))
            .unwrap_or("_");
        let params = node
            .child_by_field_name("parameters")
            .map(|n| node_text(n, source))
            .unwrap_or("()");
        let base = if ret.is_empty() {
            format!("{name}{params}")
        } else {
            format!("{ret} {name}{params}")
        };
        compact_ws(&prefixed(&mods, format_args!("{base}")))
    }

    fn field_text(&self, node: Node, source: &[u8]) -> String {
        let mods = self.modifiers_text(node, source);
        let ty = node
            .child_by_field_name("type")
            .map(|n| node_text(n, source))
            .unwrap_or("_");
        let name = find_child(node, "variable_declarator")
            .and_then(|n| n.child_by_field_name("name"))
            .map(|n| node_text(n, source))
            .unwrap_or("_");
        prefixed(&mods, format_args!("{ty} {name}"))
    }

    fn extract_interface(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let mods = self.modifiers_text(node, source);
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))?;
        let type_params = find_child(node, "type_parameters")
            .map(|n| node_text(n, source))
            .unwrap_or("");
        let extends = find_child(node, "extends_interfaces")
            .map(|n| format!(" extends {}", self.type_list_text(n, source)))
            .unwrap_or_default();

        let label = prefixed(
            &mods,
            format_args!("interface {name}{type_params}{extends}"),
        );

        let children = self.extract_interface_body(node, source);
        Some(SkeletonEntry::new(Section::Trait, node, label).with_children(children))
    }

    fn extract_interface_body(&self, node: Node, source: &[u8]) -> Vec<String> {
        let Some(body) = node.child_by_field_name("body") else {
            return Vec::new();
        };
        let mut members = Vec::new();
        let mut cursor = body.walk();
        for child in body.children(&mut cursor) {
            let text = match child.kind() {
                "method_declaration" => self.method_signature(child, source),
                "constant_declaration" => self.field_text(child, source),
                _ => continue,
            };
            let lr = line_range(child.start_position().row + 1, child.end_position().row + 1);
            members.push(format!("{text} {lr}"));
        }
        members
    }

    fn extract_enum(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let mods = self.modifiers_text(node, source);
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))?;
        let type_params = find_child(node, "type_parameters")
            .map(|n| node_text(n, source))
            .unwrap_or("");
        let interfaces = self.implements_clause(node, source);
        let label = prefixed(&mods, format_args!("enum {name}{type_params}{interfaces}"));

        let body = node.child_by_field_name("body")?;
        let mut constants = Vec::new();
        let mut cursor = body.walk();
        for child in body.children(&mut cursor) {
            if child.kind() == "enum_constant" {
                let cname = child
                    .child_by_field_name("name")
                    .map(|n| node_text(n, source))
                    .unwrap_or("_");
                constants.push(cname.to_string());
            }
        }

        Some(
            SkeletonEntry::new(Section::Type, node, label)
                .with_children(constants)
                .with_child_kind(ChildKind::Brief),
        )
    }

    fn extract_record(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let mods = self.modifiers_text(node, source);
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))?;
        let type_params = find_child(node, "type_parameters")
            .map(|n| node_text(n, source))
            .unwrap_or("");
        let params = find_child(node, "formal_parameters")
            .map(|n| node_text(n, source))
            .unwrap_or("()");

        let interfaces = self.implements_clause(node, source);
        let label = prefixed(
            &mods,
            format_args!("record {name}{type_params}{params}{interfaces}"),
        );

        Some(SkeletonEntry::new(Section::Type, node, label))
    }

    fn extract_annotation_type(&self, node: Node, source: &[u8]) -> Option<SkeletonEntry> {
        let mods = self.modifiers_text(node, source);
        let name = node
            .child_by_field_name("name")
            .map(|n| node_text(n, source))?;
        let label = prefixed(&mods, format_args!("@interface {name}"));
        Some(SkeletonEntry::new(Section::Type, node, label))
    }
}

impl LanguageExtractor for JavaExtractor {
    fn extract_nodes(&self, node: Node, source: &[u8], _attrs: &[Node]) -> Vec<SkeletonEntry> {
        match node.kind() {
            "import_declaration" => self.extract_import(node, source).into_iter().collect(),
            "package_declaration" => self.extract_package(node, source).into_iter().collect(),
            "class_declaration" => self.extract_class(node, source).into_iter().collect(),
            "interface_declaration" => self.extract_interface(node, source).into_iter().collect(),
            "enum_declaration" => self.extract_enum(node, source).into_iter().collect(),
            "record_declaration" => self.extract_record(node, source).into_iter().collect(),
            "annotation_type_declaration" => self
                .extract_annotation_type(node, source)
                .into_iter()
                .collect(),
            _ => Vec::new(),
        }
    }

    fn is_doc_comment(&self, node: Node, source: &[u8]) -> bool {
        node.kind() == "block_comment" && node_text(node, source).starts_with("/**")
    }

    fn import_separator(&self) -> &'static str {
        "."
    }
}

File: plugins/index/lang/java.lua
return function(U)
  local get_text = U.get_text
  local find_child = U.find_child
  local compact_ws = U.compact_ws
  local format_range = U.format_range
  local line_start = U.line_start
  local line_end = U.line_end
  local new_entry = U.new_entry
  local new_import_entry = U.new_import_entry
  local simple_import = U.simple_import
  local SECTION = U.SECTION
  local CHILD_BRIEF = U.CHILD_BRIEF
  local extract_enum_variants = U.extract_enum_variants
  local extract_body_members = U.extract_body_members

  local function modifiers_text(node, source)
    local mods = find_child(node, "modifiers")
    if not mods then return "" end
    local parts = {}
    for _, child in ipairs(mods:children()) do
      local ck = child:type()
      local text = get_text(child, source)
      if ck == "marker_annotation" or ck == "annotation" then
        parts[#parts + 1] = text
      elseif text:match("^public$") or text:match("^private$") or text:match("^protected$")
          or text:match("^static$") or text:match("^final$") or text:match("^abstract$")
          or text:match("^default$") or text:match("^synchronized$") then
        parts[#parts + 1] = text
      end
    end
    return table.concat(parts, " ")
  end

  local function tparams_text(node, source)
    local tp = find_child(node, "type_parameters")
    return tp and get_text(tp, source) or ""
  end

  local function method_sig(node, source)
    local mods = modifiers_text(node, source)
    local name_node = node:field("name")[1]
    if not name_node then return nil end
    local name = get_text(name_node, source)
    local ret_node = node:field("type")[1]
    local ret = ret_node and (get_text(ret_node, source) .. " ") or ""
    local params_node = node:field("parameters")[1]
    local params = params_node and get_text(params_node, source) or "()"
    return compact_ws(mods ~= "" and (mods .. " " .. ret .. name .. params) or (ret .. name .. params))
  end

  local function field_text(node, source)
    local mods = modifiers_text(node, source)
    local type_node = node:field("type")[1]
    local type_str = type_node and get_text(type_node, source) or ""
    local decl = find_child(node, "variable_declarator")
    local fname = decl and decl:field("name")[1]
    local fname_str = fname and get_text(fname, source) or "_"
    return compact_ws(mods ~= "" and (mods .. " " .. type_str .. " " .. fname_str) or (type_str .. " " .. fname_str))
  end

  local CLASS_RULES = {
    { kind = "method_declaration",      handler = "method", fn = method_sig },
    { kind = "constructor_declaration", handler = "method", fn = method_sig },
    { kind = "field_declaration",       handler = "field",  fn = field_text, counter = "field" },
  }

  local INTERFACE_RULES = {
    { kind = "method_declaration",   handler = "method", fn = method_sig },
    { kind = "constant_declaration", handler = "field",  fn = field_text, counter = "field" },
  }

  local function type_list_text(parent, source)
    local tl = find_child(parent, "type_list")
    if tl then return get_text(tl, source) end
    local raw = get_text(parent, source)
    return raw:gsub("^extends%s*", ""):gsub("^implements%s*", "")
  end

  local function implements_clause(node, source)
    local ifaces = node:field("interfaces")[1]
    if not ifaces then return "" end
    return " implements " .. type_list_text(ifaces, source)
  end

  local function extract_class(node, source)
    local mods = modifiers_text(node, source)
    local name_node = node:field("name")[1]
    if not name_node then return nil end
    local name = get_text(name_node, source)
    local tparams = tparams_text(node, source)
    local super_node = node:field("superclass")[1]
    local super = ""
    if super_node then
      local type_id = find_child(super_node, "type_identifier")
      super = " extends " .. get_text(type_id or super_node, source)
    end
    local interfaces = implements_clause(node, source)
    local label = compact_ws(mods ~= "" and (mods .. " class " .. name .. tparams .. super .. interfaces)
                             or ("class " .. name .. tparams .. super .. interfaces))
    local entry = new_entry(SECTION.Class, node, label)
    local body = find_child(node, "class_body")
    if body then entry.children = extract_body_members(body, source, CLASS_RULES) end
    return entry
  end

  local function extract_interface(node, source)
    local mods = modifiers_text(node, source)
    local name_node = node:field("name")[1]
    if not name_node then return nil end
    local name = get_text(name_node, source)
    local tparams = tparams_text(node, source)
    local ext_node = find_child(node, "extends_interfaces")
    local ext_str = ext_node and (" extends " .. type_list_text(ext_node, source)) or ""
    local label = compact_ws(mods ~= "" and (mods .. " interface " .. name .. tparams .. ext_str)
                             or ("interface " .. name .. tparams .. ext_str))
    local entry = new_entry(SECTION.Trait, node, label)
    local body = find_child(node, "interface_body")
    if body then entry.children = extract_body_members(body, source, INTERFACE_RULES) end
    return entry
  end

  local function extract_enum(node, source)
    local mods = modifiers_text(node, source)
    local name_node = node:field("name")[1]
    if not name_node then return nil end
    local name = get_text(name_node, source)
    local tparams = tparams_text(node, source)
    local interfaces = implements_clause(node, source)
    local label = compact_ws(mods ~= "" and (mods .. " enum " .. name .. tparams .. interfaces)
                             or ("enum " .. name .. tparams .. interfaces))
    local entry = new_entry(SECTION.Type, node, label)
    local body = find_child(node, "enum_body")
    if body then
      entry.children = extract_enum_variants(body, source, "enum_constant")
      entry.child_kind = CHILD_BRIEF
    end
    return entry
  end

  local function extract_record(node, source)
    local mods = modifiers_text(node, source)
    local name_node = node:field("name")[1]
    if not name_node then return nil end
    local name = get_text(name_node, source)
    local tparams = tparams_text(node, source)
    local params_node = find_child(node, "formal_parameters")
    local params = params_node and get_text(params_node, source) or ""
    local interfaces = implements_clause(node, source)
    local label = compact_ws(mods ~= "" and (mods .. " record " .. name .. tparams .. params .. interfaces)
                             or ("record " .. name .. tparams .. params .. interfaces))
    return new_entry(SECTION.Class, node, label)
  end

  local function extract_annotation_type(node, source)
    local mods = modifiers_text(node, source)
    local name_node = node:field("name")[1]
    if not name_node then return nil end
    local name = get_text(name_node, source)
    local label = compact_ws(mods ~= "" and (mods .. " @interface " .. name) or ("@interface " .. name))
    return new_entry(SECTION.Type, node, label)
  end

  local function extract_package(node, source)
    local text = get_text(node, source)
    local cleaned = text:match("^package%s+(.-)%s*;?%s*$") or text
    return new_entry(SECTION.Module, node, cleaned)
  end

  return {
    import_separator = ".",

    is_doc_comment = function(node, source)
      if node:type() ~= "block_comment" then return false end
      return get_text(node, source):sub(1, 3) == "/**"
    end,

    extract_nodes = function(node, source, _attrs)
      local kind = node:type()
      if kind == "import_declaration" then
        return { simple_import(node, source, { "import%s+" }, ".") }
      elseif kind == "package_declaration" then
        return { extract_package(node, source) }
      elseif kind == "class_declaration" then
        local e = extract_class(node, source)
        return e and { e } or {}
      elseif kind == "interface_declaration" then
        local e = extract_interface(node, source)
        return e and { e } or {}
      elseif kind == "enum_declaration" then
        local e = extract_enum(node, source)
        return e and { e } or {}
      elseif kind == "record_declaration" then
        local e = extract_record(node, source)
        return e and { e } or {}
      elseif kind == "annotation_type_declaration" then
        local e = extract_annotation_type(node, source)
        return e and { e } or {}
      end
      return {}
    end,
  }
end
